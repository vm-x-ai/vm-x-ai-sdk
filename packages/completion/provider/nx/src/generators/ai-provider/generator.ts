import * as path from 'path';
import type { Tree } from '@nx/devkit';
import {
  readProjectConfiguration,
  updateProjectConfiguration,
  formatFiles,
  generateFiles,
  addDependenciesToPackageJson,
  updateJson,
} from '@nx/devkit';
import { libraryGenerator } from '@nx/js';
import { getPackageVersion } from '@vm-x-ai/completion-provider';
import type { AiProviderGeneratorSchema } from './schema';

export async function aiProviderGenerator(tree: Tree, options: AiProviderGeneratorSchema) {
  const callback = await libraryGenerator(tree, {
    ...options,
    publishable: false,
  });

  const projectConfig = readProjectConfiguration(tree, options.name);
  tree.delete(path.join(projectConfig.root, 'src/lib', `${options.name}.ts`));
  tree.delete(path.join(projectConfig.root, 'src/lib', `${options.name}.spec.ts`));

  updateJson(tree, path.join(projectConfig.root, 'tsconfig.json'), (json) => {
    delete json.compilerOptions.noPropertyAccessFromIndexSignature;
    return json;
  });

  generateFiles(tree, path.join(__dirname, 'files'), projectConfig.root, {
    ...options,
    dot: '.',
    title: options.name
      .split('-')
      .map((s) => s.charAt(0).toUpperCase() + s.slice(1))
      .join(' '),
  });

  const dependenciesCallback = addDependenciesToPackageJson(
    tree,
    {
      '@nestjs/common': '^10.3.9',
      'nestjs-otel': '6.1.1',
      rxjs: '^7.8.1',
      uuid: '^9.0.1',
      '@vm-x-ai/completion-provider': `^${getPackageVersion()}`,
    },
    {
      'esbuild-plugin-tsc': '^0.4.0',
      '@types/uuid': '^10.0.0',
    },
    undefined,
    true,
  );

  updateProjectConfiguration(tree, options.name, {
    ...projectConfig,
    targets: {
      ...projectConfig.targets,
      build: {
        executor: 'nx:run-commands',
        options: {
          command: 'vm-x-completion-ai-provider publish --dry-run',
          cwd: '{projectRoot}',
        },
      },
      publish: {
        executor: 'nx:run-commands',
        options: {
          command: 'vm-x-completion-ai-provider publish',
          cwd: '{projectRoot}',
        },
      },
      watch: {
        executor: 'nx:run-commands',
        options: {
          command: 'vm-x-completion-ai-provider publish --watch',
          cwd: '{projectRoot}',
        },
      },
    },
  });

  await formatFiles(tree);

  return async () => {
    await callback();
    await dependenciesCallback();
  };
}

export default aiProviderGenerator;
