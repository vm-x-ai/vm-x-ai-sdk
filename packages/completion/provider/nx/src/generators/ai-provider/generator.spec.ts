import type { Tree } from '@nx/devkit';
import { readProjectConfiguration } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { aiProviderGenerator } from './generator';
import type { AiProviderGeneratorSchema } from './schema';

describe('ai-provider generator', () => {
  let tree: Tree;
  const options: AiProviderGeneratorSchema = {
    name: 'dummy',
    directory: 'packages/dummy',
    providerType: 'community',
    providerVisibility: 'private',
  };

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  it('should run successfully', async () => {
    await aiProviderGenerator(tree, options);
    const config = readProjectConfiguration(tree, 'dummy');
    expect(config).toMatchSnapshot();
    expect(tree.listChanges().map((change) => change.path)).toMatchSnapshot();
  });
});
