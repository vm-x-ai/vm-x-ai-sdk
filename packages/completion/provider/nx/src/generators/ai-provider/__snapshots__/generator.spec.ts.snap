// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ai-provider generator > should run successfully 1`] = `
{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "name": "dummy",
  "projectType": "library",
  "root": "packages/dummy",
  "sourceRoot": "packages/dummy/src",
  "tags": [],
  "targets": {
    "build": {
      "executor": "nx:run-commands",
      "options": {
        "command": "vm-x-completion-ai-provider publish --dry-run",
        "cwd": "{projectRoot}",
      },
    },
    "lint": {
      "executor": "@nx/eslint:lint",
    },
    "publish": {
      "executor": "nx:run-commands",
      "options": {
        "command": "vm-x-completion-ai-provider publish",
        "cwd": "{projectRoot}",
      },
    },
    "test": {
      "executor": "@nx/jest:jest",
      "options": {
        "jestConfig": "packages/dummy/jest.config.ts",
      },
      "outputs": [
        "{workspaceRoot}/coverage/{projectRoot}",
      ],
    },
    "watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "vm-x-completion-ai-provider publish --watch",
        "cwd": "{projectRoot}",
      },
    },
  },
}
`;

exports[`ai-provider generator > should run successfully 2`] = `
[
  ".prettierrc",
  "package.json",
  "nx.json",
  "tsconfig.base.json",
  ".prettierignore",
  "packages/dummy/tsconfig.lib.json",
  "packages/dummy/tsconfig.json",
  "packages/dummy/src/index.ts",
  "packages/dummy/README.md",
  "packages/dummy/package.json",
  "packages/dummy/project.json",
  "eslint.config.js",
  "packages/dummy/eslint.config.js",
  "jest.preset.js",
  "jest.config.ts",
  "packages/dummy/jest.config.ts",
  "packages/dummy/tsconfig.spec.json",
  "packages/dummy/.env.local",
  "packages/dummy/assets/logos/vmx.png",
  "packages/dummy/esbuild.config.js",
  "packages/dummy/manifest.yaml",
  "packages/dummy/src/provider.spec.ts",
  "packages/dummy/src/provider.ts",
]
`;
